### ISSUE TITLE
Allow creating an AHashMap without an `Eq` or `Hash` bound on the keys

### ISSUE THREAD

**Issue by jyn514** [tkaitchuck/aHash#52]
`with_capacity_and_hasher` requires `K` to implement `Eq`:
```
error[E0277]: the trait bound `K: Hash` is not satisfied
  --> compiler/rustc_data_structures/src/sso/map.rs:89:29
   |
89 |             SsoHashMap::Map(FxHashMap::with_capacity_and_hasher(cap, Default::default()))
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Hash` is not implemented for `K`
   |
   = note: required by `AHashMap::<K, V, S>::with_capacity_and_hasher`
```
Consider relaxing that bound, I think you shouldn't need it for constructing the hashmap, only for operations involving the elements (get/set/entry). Same request for various other functions, new(), capacity(), len(), etc.

Note the error is misleading here - `FxHashMap` is an alias for `AHashMap` that I renamed because it was simpler than changing a hundred imports.

_Originally posted by @jyn514 in https://github.com/rust-lang/rust/pull/77996#issuecomment-709652697_

**Comment by tkaitchuck** [comment]
Done https://github.com/tkaitchuck/aHash/commit/00d7e747ab8928f6374fa5ae28f6435920e8aaf3

**Comment by jyn514** [comment]
Sorry, my request wasn't clear: I said 'Eq' but I meant 'neither Eq not Hash'. Both aren't necessary for these functions.

**Comment by tkaitchuck** [comment]
@jyn514 What are you doing with these hashmaps that don't contain any items?!

**Comment by jyn514** [comment]
Don't ask me, ask rustc 😆 Small set optimizations! https://doc.rust-lang.org/nightly/nightly-rustc/rustc_data_structures/sso/enum.SsoHashMap.html

**Comment by tkaitchuck** [comment]
Ok. Done https://github.com/tkaitchuck/aHash/commit/23af1822b977139c4fa6de823e030c48648924d4
Released as 0.5.3
